src/transformers/pipelines/visual_question_answering.py

```diff
@@ -118,20 +131,56 @@ def preprocess(self, inputs, padding=False, truncation=False):
         model_inputs.update(image_features)
         return model_inputs
 
-    def _forward(self, model_inputs):
-        model_outputs = self.model(**model_inputs)
+    def _forward(self, model_inputs, top_k=5, **generate_kwargs):
+        if self.model_type == ModelType.GENERATIVE:
+            generate_kwargs = copy.deepcopy(generate_kwargs)
+            generate_kwargs["return_dict_in_generate"] = True
+            generate_kwargs["output_scores"] = True
+            if "num_beams" in generate_kwargs:
+                if top_k > generate_kwargs["num_beams"]:
+                    pass  # raise
+            elif top_k > 1:
+                generate_kwargs["num_beams"] = top_k
+            else:
+                # activate beam search with two beam to compute scores
+                generate_kwargs["num_beams"] = 2
+            generate_kwargs["num_return_sequences"] = top_k
+            if "max_new_tokens" not in generate_kwargs:
+                # defaulting max_new_tokens to 100
+                generate_kwargs["max_new_tokens"] = 100
+            generate_outputs = self.model.generate(**model_inputs, **generate_kwargs)
+            model_outputs = {
+                "sequences_scores": generate_outputs.sequences_scores.reshape(
+                    (model_inputs["input_ids"].shape[0], top_k)
+                ),
+                "sequences": generate_outputs.sequences.reshape((model_inputs["input_ids"].shape[0], top_k, -1)),
+            }
```

Comment:
No need to all of this, I'd just run a `generate` as done [here](https://github.com/huggingface/transformers/blob/539e2281cd97c35ef4122757f26c88f44115fa94/src/transformers/pipelines/text_generation.py#L263). No need to check things like `num_beams`, `num_return_sequences`, `output_scores`, etc.

For a generative model, it will just generate an answer so no need for any top k scores. For a classifier, the top k scores make sense as we can for instance take the top 5 predicted class

Reply:
Also the forward here should only add `**generate_kwargs`